File: backend/app/schemas/map.py
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


class MapMetaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    token: str
    location: str
    version: str
    canvas_edge: list  # [width_m, height_m]


class GeoJSONGeometry(BaseModel):
    type: str
    coordinates: Any


class GeoJSONFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    geometry: GeoJSONGeometry | None
    properties: dict[str, Any]


class GeoJSONFeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[GeoJSONFeature]


class MapLayer(str, Enum):
    drivable_area = "drivable_area"
    road_segment = "road_segment"
    road_block = "road_block"
    lane = "lane"
    lane_connector = "lane_connector"
    carpark_area = "carpark_area"
    stop_line = "stop_line"
    ped_crossing = "ped_crossing"
    walkway = "walkway"
    road_divider = "road_divider"
    lane_divider = "lane_divider"
    traffic_light = "traffic_light"

File: frontend/src/types/map.ts
// バックエンドの schemas/map.py と対応

// GET /api/v1/maps レスポンスの各エントリ（MapMetaResponse に対応）
export interface MapMeta {
  token:       string
  location:    string          // 'boston-seaport' など
  version:     string
  canvas_edge: number[]        // [width_m, height_m]
}

// マップレイヤー種別（MapLayer enum に対応）
export type MapLayer =
  | 'drivable_area'
  | 'road_segment'
  | 'road_block'
  | 'lane'
  | 'lane_connector'
  | 'carpark_area'
  | 'stop_line'
  | 'ped_crossing'
  | 'walkway'
  | 'road_divider'
  | 'lane_divider'
  | 'traffic_light'

// GeoJSON 型（GeoJSONFeatureCollection レスポンスに対応）
export interface GeoJSONGeometry {
  type:        string
  coordinates: unknown
}

export interface GeoJSONMapFeature {
  type:       'Feature'
  geometry:   GeoJSONGeometry | null
  properties: Record<string, unknown>
}

export interface GeoJSONFeatureCollection {
  type:     'FeatureCollection'
  features: GeoJSONMapFeature[]
}

// GET /api/v1/instances/{token}/best-camera レスポンス（BestCameraResponse に対応）
export interface BestCamera {
  channel:           string   // 'CAM_FRONT' など
  sample_data_token: string
}
